sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubProducerAsyncClient.java

```diff
@@ -447,18 +512,65 @@ public String getEventHubName() {
             // Start send span and store updated context
             parentContext.set(tracerProvider.startSpan(finalSharedContext, ProcessKind.SEND));
         }
-
-        return withRetry(getSendLink(batch.getPartitionId())
-            .flatMap(link ->
-                messages.size() == 1
+        if (isIdempotentPartitionPublishing) {
+            PartitionPublishingState publishingState = getClientPartitionPublishingState(batch.getPartitionId());
+            return Mono.fromRunnable(() -> {
+                publishingState.getSemaphore().acquireUninterruptibly();
+                int seqNumber = publishingState.getSequenceNumber();
+                for (EventData eventData : batch.getEvents()) {
+                    eventData.setProducerGroupIdInSysProperties(publishingState.getProducerGroupId());
```

Comment:
I give it this specific name because `setProducerGroupIdInSysProperties` is internal and it doesn't set the value of property `producerGroupId`. After `setProducerGroupIdInSysProperties` is called, `getProducerGroupId` still returns null. After the event data is successfully sent out, the `commitProducerDataFromSysProperties` is called to set the value of `producerGroupId` and two other properties. Then `getProducerGroupId` returns the `producerGroupId` with a value. So I think `setProducerGroupIdInSysProperties` would be more specific and meaningful than `setProducerGroupId` when somebody else reads the code.